simple-proofs/list-algorithms.md

```diff
@@ -0,0 +1,425 @@
+  | --LIST_SUM_BODY--------- (lam in_lst 
+  |  | --...LOOP_BODY--------- (force [ (force (builtin ifThenElse))
+  |  |  |                        [ (force (builtin nullList)) in_lst ]
+  |  |  |                        ( delay (con integer 0) )
+  |  |  |                        ( delay
+  |  |  |                          [ (builtin addInteger)
+  |  |  |                            [ (force (builtin headList)) in_lst ]
+  |  |  |                            [ f_lstSum [ (force (builtin tailList)) in_lst ] ]
+  |  |  |                          ]
+  |  |  |                        )
+  |  | ----------------------- ])
+  |  ----------------------- )
+  |                        ) 
+  ----------------------- ]
+```
+
+To express correctness of list-sum, we also have to define a predicate 
+`allInts(IS)`, which checks that a given list `IS` is a list of integers:
+
+```
+  syntax Bool ::= allInts(ConstantList) [function, functional]
+
+  rule allInts(              .ConstantList) => true                              
+  rule allInts(C:Constant, IS:ConstantList) => isInt(C) andBool allInts(IS) 
+```
+
+and a predicate, `sum(XS)`, which sums a given list of integers:
+
+```
+  syntax Int ::= sum(ConstantList) [function]
+
+  rule sum(         .ConstantList ) => 0              [simplification]
+  rule sum(I:Int, IS:ConstantList ) => I +Int sum(IS) [simplification]
+```
+
+With these predicates in place, the correctness claim for the list-sum
+algorithm is formulated as follows:
+
+```k
+  claim <k>
+    [ LIST_SUM ( con list(integer) [ XS:ConstantList ] ) ] 
+      => 
+    < con integer sum(XS) >
+    ... </k>
+    <env> _ => .Map </env>
+    requires allInts(XS)
+```
+
+stating that the list-sum algorithm terminates, returning the 
+sum of the passed-in list. As for list-free, the initial environment
+is arbitrary. and the final environment is empty. In addition, 
+we have to state that the passed-in list, `XS`, has to be a list
+of integer, using `requires allInts(XS)`.
+
+The required loop invariant is formulated as follows:
+
+```k
+  claim 
+    <k>
+      LIST_SUM_LOOP_BODY
+        =>
+      < con integer sum(XS) > ... 
+    </k>
+    <env>
+      RHO:Map 
+        [ f_lstSum <- < lam y_0 [ x_0 x_0 y_0 ] RHO_1 [ x_0 <- < lam x_0 [ f_0 ( lam y_0 [ x_0 x_0 y_0 ] ) ] RHO_1 > ] > ] 
+        [ in_lst <- < con list ( integer ) [ XS:ConstantList ] > ] => .Map 
+    </env>
+    requires RHO_1 ==K RHO [ f_0 <- < lam f_lstSum LIST_SUM_BODY RHO > ]
+      andBool allInts(XS)
+```
+
+and comparing it to the list-free invariant, we can see that the only differences are: 
+
+- `< con integer sum(XS) >` instead of `< con list(integer) [ .ConstantList ] >` (semantic);
+- `f_lstSum` instead of `f_lstFree` (syntactic);
+- `LIST_SUM_LOOP_BODY` instead of `LIST_FREE_LOOP_BODY` (syntactic); and
+- `LIST_SUM_BODY` instead of `LIST_FREE_BODY` (syntactic).
+
+## List-length
+
+The next algorithm that we consider is the list-sum algorithm, 
```

Comment:
I think there was a copy-and-paste here that was not completely refined: it should be list-length and not list-sum in this section.

Reply:
To say it was not completely refined would be quite an understatement. Should be fixed now.